src/classes/EP_TaskDependencyCheck_TDTM.cls

```diff
@@ -0,0 +1,161 @@
+    * @return dmlWrapper.  
+    ********************************************************************************************************/
+    public override DmlWrapper run(List<SObject> listNew, List<SObject> listOld, 
+        TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+        /*********************************************************************************************************
+        * @description A list of the tasks that require validation.
+        */
+        List<Engagement_Plan_Task__c> listToValidate = new List<Engagement_Plan_Task__c>();
+
+        // AFTER INSERT
+        if (triggerAction == TDTM_Runnable.Action.AfterInsert) {
+            for (SObject newSO : listNew) {
+                Engagement_Plan_Task__c newEPTask = (Engagement_Plan_Task__c)newSO;
+                if (newEPTask.Dependent_Task__c != null) {
+                    listToValidate.add(newEPTask);
+                }
+            }
+            if (listToValidate.size() > 0) {
+                validateTasks(listToValidate);
+            }
+        }
+
+        // AFTER UPDATE
+        if (triggerAction == TDTM_Runnable.Action.AfterUpdate) {
+            Map<Id, Engagement_Plan_Task__c> mapOld = new Map<Id, Engagement_Plan_Task__c>((List<Engagement_Plan_Task__c>)listOld);
+
+            for (SObject newSO : listNew) {
+                Engagement_Plan_Task__c newEPTask = (Engagement_Plan_Task__c)newSO;
+                Engagement_Plan_Task__c oldEPTask = mapOld.get(newEPTask.Id);
+
+                if ((newEPTask.Dependent_Task__c != null
+                    && newEPTask.Dependent_Task__c != oldEPTask.Dependent_Task__c)
+                        || newEPTask.Engagement_Plan_Template__c != oldEPTask.Engagement_Plan_Template__c) {
+                        listToValidate.add(newEPTask);
+                }
+            }
+            if (listToValidate.size() > 0) {
+                validateTasks(listToValidate);
+            }
+        }
+
+        return null;
+    }
+
+    /*******************************************************************************************************
+    * @description For the Tasks that are new or where the EP Template or Dependency has changed,
+    * prevent cross dependencies and mismatched templates
+    * @param listPSCNew the list of PSCs from trigger new. 
+    * @param listPSCOld the list of PSCs from trigger old.
+    * @return void.  
+    ********************************************************************************************************/
+    public static void validateTasks(List<Engagement_Plan_Task__c> listToValidate) {
+        // A set of the Engagement Plan Template Ids related to the tasks.
+        Set<Id> templateIds = new Set<Id>();        
+        // Engagement Plan Task mapped to the template and all the template's related tasks
+        Map<Id, Engagement_Plan_Template__c> mapTaskIdToTemplate = new Map<Id, Engagement_Plan_Template__c>();
```

Comment:
I would make this a map from the template id to the template object, since each task has a master-detail to the template, it will end up holding less objects in memory and be easier to grock.